Issue 2296. Adding support for UsePreviousValue.

aronkyle-lightspeedretail opened the pull request:
Adding support for cloudformation UsePreviousValue parameter.

In response to issue #2296.

kyleknap commented on tests/integration/cloudformation/test_connection.py:
```diff
@@ -114,9 +131,37 @@ def test_large_template_stack_size(self):
+        self.assertEquals([('Parameter1', 'initial_value'),
+                           ('Parameter2', 'updated_value')], params)
+        
+        for _ in range(30):
```
What is the reason for sleeping again when nothing else is happening after other than cleaning up the stack by deleting it?

aronkyle-lightspeedretail replied:
It's been quite a few months but I think it had to do with stack.update() responding immediately with a status of IN PROGRESS. So we need some kind of delay to keep the test going long enough for the update to actually take affect before attempting the update with the "previous value" in a second update. 

I'm open suggestions if there's a better way.

kyleknap replied:
But isn't that what you do at this wait (which I am fine with): https://github.com/boto/boto/pull/3029/files#diff-e70f6fdc16bd22f32c84d3977d1d9f5cR142

I was just confused why you wait again after you test that the stack successfully has been updated such that it uses both the new and the previous values.

aronkyle-lightspeedretail replied:
Oh I see. I think that is just to guarantee that cloud formation is in a "good" state when the test completes. e.g. a tear down could potentially fail if the stack was still updating.

If you want that out I can remove as soon as a find some time.

kyleknap replied:
Cool. Yeah I would prefer if you remove it.

kyleknap commented on tests/unit/cloudformation/test_connection.py:
```diff
@@ -149,6 +150,11 @@ def test_update_stack_all_args(self):
             'Parameters.member.1.ParameterKey': 'KeyName',
             'Parameters.member.1.ParameterValue': 'myKeyName',
+            'Parameters.member.2.ParameterKey': 'KeyName2',
+            'Parameters.member.2.ParameterValue': '',
```
You may need to update the logic to exclude ``ParameterValue`` if ``UsePreviousValue`` is being used. It looks like it works according to the integration test, but I would like to stick with what the API references say, which is to exclude a parameter value if specify UsePreviousValue is true: http://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_Parameter.html

aronkyle-lightspeedretail replied:
Again, it's been a few months. But I don't think the API  will accept it otherwise.

aronkyle-lightspeedretail replied:
I'll try and find some time to remove and test the above change. If I can update the logic I will. Otherwise we might have to leave it as is.

kyleknap replied:
Thanks! I think we will need to do the update to the logic.

I just tested the integration tests where you set the previous value to true:
https://github.com/boto/boto/pull/3029/files#diff-e70f6fdc16bd22f32c84d3977d1d9f5cR151

It looks like this passes because the value in the second element of the tuple is ``''``. If you put a value like ``'foo'`` in as the second element, you get the following error:
```
boto: ERROR: 400 Bad Request
boto: ERROR: {"Error":{"Code":"ValidationError","Message":"Invalid input for parameter key Parameter1. Cannot specify usePreviousValue as true and non empty value for a parameter","Type":"Sender"},"RequestId":"b6c9aeeb-cdab-11e4-9e3d-df34bd1ee4bf"}
```

Don't worry, the fix is easy to make the ``'foo'`` test case work.

Just make the logic in ``connection.py``:
```
params['Parameters.member.%d.ParameterKey' % (i + 1)] = key
if use_previous:
    params['Parameters.member.%d.UsePreviousValue' 
                 % (i + 1)] = self.encode_bool(use_previous)
else:
    params['Parameters.member.%d.ParameterValue' % (i + 1)] = value

```
I tested the integration test and that worked for me.

kyleknap commented:
@aronkyle-lightspeedretail 
Thanks for the pull request! It looks good to me in general. I just had a couple comments that I would like you to address before I merge it.

aronkyle-lightspeedretail commented:
@kyleknap 

I updated the call to omit the parameter value if user previous value is true.

I also removed the extra wait loop at the end of the integration test.

That said, if you run just the one test, the test passed, but an error is returned trying to delete the stack since it hasn't finished updating yet.

kyleknap commented:
@aronkyle-lightspeedretail 

Oh, if an error is returned when you try to delete the stack, then we should keep the extra wait loop. It is more important that resources are cleaned up. Thanks for hanging in there with me.

aronkyle-lightspeedretail commented:
@kyleknap The wait loop is back in. Integration test passes.

kyleknap commented:
Thanks for seeing the PR through! Merging.

aronkyle-lightspeedretail commented:
Cheers. ;)
